lumen/ai/tools.py

```diff
@@ -104,45 +104,100 @@ class TableLookup(VectorLookupTool):
+        self._table_metadata[table_name] = metadata
+
+        vector_metadata = {"table": table, "source": source.name}
+        if self.vector_store.filter_by(vector_metadata):
+            return
+
+        if description := metadata.get('description', ''):
+            enriched_text += f"\nDescription: {description}"
+        if self.include_columns and (columns := metadata.get('columns', {})):
+            enriched_text += "\nColumns:"
+            for col_name, col_info in columns.items():
+                col_text = f"\n- {col_name}"
+                if 'description' in col_info:
+                    col_text += f": {col_info['description']}"
+                enriched_text += col_text
+        self.vector_store.add([{"text": enriched_text, "metadata": vector_metadata}])
+
     def _update_vector_store(self, _, __, sources):
         for source in sources:
             for table in source.get_tables():
-                source_table = f'{source.name}{SOURCE_TABLE_SEPARATOR}{table}'
-                if not self.vector_store.query(source_table, threshold=1):
-                    self.vector_store.add([{"text": source_table}])
+                metadata = {"table": table, "source": source.name}
+                if not self.include_metadata:
+                    if self.vector_store.filter_by(metadata):
+                        # TODO: Add ability to update existing source table?
+                        continue
+                    source_table = f"{source.name}{SOURCE_TABLE_SEPARATOR}{table}"
+                    self.vector_store.add([{"text": source_table, "metadata": metadata}])
+                else:
+                    task = asyncio.create_task(
+                        self._fetch_and_store_metadata(source, table)
+                    )
+                    self._metadata_tasks.add(task)
+                    task.add_done_callback(lambda t: self._metadata_tasks.discard(t))
 
     async def respond(self, messages: list[Message], **kwargs: dict[str, Any]) -> str:
-        closest_tables = []
-        message = "The most relevant tables are:\n"
-        results = self.vector_store.query(
-            messages[-1]["content"],
-            top_k=self.n,
-            threshold=self.min_similarity,
-        )
+        """
+        Fetches the closest tables based on the user query, depending on whether
+        any tables reach the min_similarity threshold we adjust the messsage.
+        """
+        query = messages[-1]["content"]
+        results = self.vector_store.query(query, top_k=self.n)
+        multi_source = len(self._memory.get("sources", []))
+        any_matches = any(result['similarity'] >= self.min_similarity for result in results)
+        closest_tables, descriptions = [], []
         for result in results:
-            if SOURCE_TABLE_SEPARATOR not in result["text"]:
-                closest_tables.append(result["text"])
+            if result['similarity'] < self.min_similarity and any_matches:
```

Comment:
I have a hard time figuring out why any_matches is appended, rather than prepended.

Reply:
Doesn't matter much, but sure I'll move it first.